graveler/graveler.go

```diff
@@ -448,6 +446,16 @@ type StagingManager interface {
 	DropByPrefix(ctx context.Context, st StagingToken, prefix Key) error
 }
 
+// BranchLocker enforces the branch locking logic
+// The logic is as follows:
+// - Allow concurrent writers to acquire the lock.
+// - A Metadata update waits for all current writers to release the lock, and then gets the lock.
+// - While a metadata update has the lock or is waiting for the lock, any other operation fails to acquire the lock.
+type BranchLocker interface {
+	Writer(ctx context.Context, repositoryID RepositoryID, branchID BranchID, lockedCB func() (interface{}, error)) (interface{}, error)
```

Comment:
lockedCB could be at type alias, kind of like `context.CancelFunc`, to avoid repetition...

Reply:
done - a little type coupling, but nice to read.